File: ArbolBinarioJs/ABB.js
class Nodo {
  constructor(id) {
    this.id = id;
    this.izquierda = null;
    this.derecha = null;
  }
}

class ArbolABB {
  constructor() {
    this.raiz = null;
  }

  estavacia() {
    return this.raiz === null;
  }

  agregar(id, node = this.raiz) {
    if (!node) {
      this.raiz = new Nodo(id);
      return;
    }

    if (id < node.id) {
      if (node.izquierda) {
        return this.agregar(id, node.izquierda);
      }
      node.izquierda = new Nodo(id);
      return;
    } else {
      if (node.derecha) {
        return this.agregar(id, node.derecha);
      }
      node.derecha = new Nodo(id);
      return;
    }
  }

  preOrden(node = this.raiz) {
    if(!node){
      return;
    }
    console.log(node.id);
    this.preOrden(node.izquierda)
    this.preOrden(node.derecha)
  }

  postOrden(node = this.root) {
    if(!node){
      return;
    }
    this.postOrden(node.izquierda);
    this.postOrden(node.derecha);
    console.log(node.id);
  }

  InOrden(node = this.raiz) {
    if(!node){
      return;
    }
    this.InOrden(node.izquierda);
    console.log(node.id);
    this.InOrden(node.derecha);
  }
}

arbolito = new ArbolABB();
arbolito.agregar(15);
arbolito.agregar(5);
arbolito.agregar(3);
arbolito.agregar(12);
arbolito.agregar(10);
arbolito.agregar(13);
arbolito.agregar(6);
arbolito.agregar(7);
arbolito.agregar(16);
arbolito.agregar(20);
arbolito.agregar(18);
arbolito.agregar(23);

arbolito.preOrden(arbolito.raiz);
arbolito.postOrden(arbolito.raiz);
arbolito.InOrden(arbolito.raiz);

File: arboles/arbolBinarioPython/ABB.py
import os
from Nodo import Nodo

class ArbolABB():
    def __init__(self):
        self.raiz = None
    
    def estaVacio(self):
        return self.raiz == None
    
    def agregar(self, id):
        self.raiz = self.__agregarInterno(id, self.raiz)
    
    def __agregarInterno(self, id, raiz):
        if raiz==None:
            return Nodo(id)
        else:
            if id< raiz.id:
                raiz.izquierda = self.__agregarInterno(id, raiz.izquierda)
            elif id>raiz.id:
                raiz.derecha = self.__agregarInterno(id, raiz.derecha)
        return raiz

    def preOrden(self, root):
        self.__preOrdenInterno(root)
    
    def __preOrdenInterno(self, root):
        if root != None:
            print(end="| ")
            self.__preOrdenInterno(root.izquierda)
            print(root.id, end=" | ")
            self.__preOrdenInterno(root.derecha)
    
    def postOrden(self, root):
        self.__postOrdenInterno(root)
    
    def __postOrdenInterno(self, root):
        if root != None:
            print(end="| ")
            self.__postOrdenInterno(root.derecha)
            print(root.id, end=" | ")
            self.__postOrdenInterno(root.izquierda)
    
    def InOrden(self, root):
        self.__inOrdenInterno(root)

    def __inOrdenInterno(self, root):
        if root!=None:
            print(end="| ")
            print(root.id, end=" | ")
            self.__inOrdenInterno(root.derecha)
            self.__inOrdenInterno(root.izquierda)
    
    def graficar(self, raiz):
        file = open("Arbol.dot", 'w')
        file.write("digraph G { \n")
        file.write("rankdir=TB; \n")
        file.write("node [shape = record, color=black , style=filled, fillcolor=gray93];\n")        
        file.write(self.__graficadora(raiz))
        file.write("} \n")
        file.close()
        os.system("dot -Tpng Arbol.dot -o Arbol.png")
        os.startfile("Arbol.png")
    
    def __graficadora(self, root):
        cadena =""
        if(root.derecha == None) and (root.izquierda == None):
            cadena = "nodo"+str(root.id)+"[label =\" "+str(root.id) + "\"]; \n"
        else:
            cadena = "nodo"+str(root.id)+"[label =\"<C0>| "+str(root.id) + "|<C1> \"]; \n"

        if(root.izquierda!=None):
            cadena = cadena + self.__graficadora(root.izquierda)+"nodo"+str(root.id)+":C0->nodo"+str(root.izquierda.id)+"\n"
        if(root.derecha!=None):
            cadena = cadena + self.__graficadora(root.derecha)+"nodo"+str(root.id)+":C1->nodo"+str(root.derecha.id)+"\n"
        
        return cadena